common/src/chain/block/block_v1.rs

```diff
@@ -20,7 +27,42 @@ pub struct BlockV1 {
     pub(super) transactions: Vec<Transaction>,
 }
 
+impl BlockVersion for BlockV1 {
+    const BLOCK_VERSION: u32 = 1;
+}
+
+impl Idable<BlockHeader> for BlockHeader {
+    fn get_id(&self) -> Id<Self> {
+        Id::new(&id::hash_encoded(self))
+    }
+}
```

Comment:
It is somewhat suboptimal IMHO for `block_header.get_id()` to return `Id<BlockHeader>` rather than `Id<Block>`, since it is essentially the same thing. We can decouple the annotation from the type by changing the `Idable` trait:
```rust
trait Idable: ?Sized {
    type Tag: ?Sized;
    fn get_id(&self) -> Id<Self::Tag>;
}
```

The `Tag` type would be almost always the same as `Self` but it would allow us to return `Id<Block>` form `block_header.get_id()`. It would also allow us to provide implementations for `Idable` on various wrapper types like `Box<T>` with result `Id<T>` rather than `Id<Box<T>>`.

Reply:
Also I noticed `Idable` currently has a superfluous type argument
```rust
pub trait Idable<T: ?Sized>
```
should be:
```rust
pub trait Idable
```

Reply:
I think a better solution is to keep `T` and use it as the tag. I'll push a commit soon that does that.

Reply:
> I think a better solution is to keep `T` and use it as the tag. I'll push a commit soon that does that.

That is also a possibility. Just keep in mind there's a difference between using `T` as tag and having Tag associated type as I presented above. With `Idable<T>` you can have *multiple* `Idable` implementations on each type. E.g. you can have `Idable<Foo> for Block` and `Idable<Bar> for Block`, giving you `Id<Foo>` and `Id<Bar>`, respectively. With associated type, there is one canonical `Idable` for each type implementing it.

Reply:
Alrighty. I'll go for the tag. It fits better.